app/grandchallenge/publications/utils/__init__.py

```diff
@@ -1,11 +1,75 @@
+import re
+
 import requests
+from django.core.validators import RegexValidator
+from django.db import models
+from django.forms import CharField
 
-from grandchallenge.publications.models import (
-    PublicationType,
-    get_publication_type,
-)
 from grandchallenge.publications.utils.manubot import get_arxiv_csl
 
+# regex modified for python syntax from
+# https://www.crossref.org/blog/dois-and-matching-regular-expressions/
+DOI_REGEX = r"^10\.\d{4,9}/[-._;()/:a-z0-9]+$"
+ARXIV_REGEX = r"^\d{4}\.\d{4,5}$"
+
+identifier_validator = RegexValidator(regex=f"{DOI_REGEX}|{ARXIV_REGEX}")
+
+
+class PublicationType(models.TextChoices):
+    DOI = "D"
+    ARXIV = "A"
+
+
+class IdentifierField(models.CharField):
+
+    description = (
+        "The DOI e.g., 10.1002/mrm.25227, or the arXiv id, e.g., 2006.12449"
+    )
+
+    def __init__(self, *args, **kwargs):
+        kwargs["max_length"] = 255
+        kwargs[
+            "help_text"
+        ] = "The DOI, e.g., 10.1002/mrm.25227, or the arXiv id, e.g., 2006.12449"
+        kwargs["validators"] = [identifier_validator]
+        super().__init__(*args, **kwargs)
+
+    def deconstruct(self):
+        name, path, args, kwargs = super().deconstruct()
+        del kwargs["max_length"]
+        del kwargs["help_text"]
+        del kwargs["validators"]
```

Comment:
Does this mean that users cannot change the help_text or validators? I think we would still want to keep that as an option.

Reply:
Yes, indeed, as it is now the `help_text` and the `validators` cannot be changed by users. I agree that it makes sense to leave the `help_text` customizable, but why would we want the `validators` to be customizable? Isn't the validation **the** thing that defines this custom field?

Reply:
Not really - we're using a custom field here to add helper methods to make the field more usable (see my first review comment). You want to force that this validator is called, but you also want the user to be able to extend the validation used (say, I only want arxiv ids). The way to do this is not to force the validators to be only this one, but to either extend the list of validators so the one you want is always there, or to extend the fields `validate` method:

> The validate() method on a Field handles field-specific validation that is not suitable for a validator. It takes a value that has been coerced to a correct datatype and raises ValidationError on any error. This method does not return anything and shouldn’t alter the value. You should override it to handle validation logic that you can’t or don’t want to put in a validator.
https://docs.djangoproject.com/en/4.0/ref/forms/validation/

e.g. https://github.com/django/django/blob/27b07a3246bc033cd9ded01238c6dc64731cce35/django/db/models/fields/json.py#L104-L113.

This way the fields special validation is always done, but the consumer can do additional things.

The same would go for the max_length field, you can set a default but this forcing pattern makes the field more difficult to use.

Reply:
Here is a good example with a default validator and max_length preset https://github.com/django/django/blob/27b07a3246bc033cd9ded01238c6dc64731cce35/django/db/models/fields/__init__.py#L2523-L2545